diagnose.go

```diff
@@ -0,0 +1,693 @@
+			if i > 0 {
+				if err := di.writeString(", "); err != nil {
+					return err
+				}
+			}
+
+			i++
+			if err := di.item(); err != nil {
+				return err
+			}
+		}
+
+		return di.writeByte(')')
+
+	case 0x9f: // indefinite array
+		di.d.off++
+		if err := di.writeString("[_ "); err != nil {
+			return err
+		}
+
+		i := 0
+		for !di.d.foundBreak() {
+			if i > 0 {
+				if err := di.writeString(", "); err != nil {
+					return err
+				}
+			}
+
+			i++
+			if err := di.item(); err != nil {
+				return err
+			}
+		}
+
+		return di.writeByte(']')
+
+	case 0xbf: // indefinite map
+		di.d.off++
+		if err := di.writeString("{_ "); err != nil {
+			return err
+		}
+
+		i := 0
+		for !di.d.foundBreak() {
+			if i > 0 {
+				if err := di.writeString(", "); err != nil {
+					return err
+				}
+			}
+
+			i++
+			// key
+			if err := di.item(); err != nil {
+				return err
+			}
+
+			if err := di.writeString(": "); err != nil {
+				return err
+			}
+
+			// value
+			if err := di.item(); err != nil {
+				return err
+			}
+		}
+
+		return di.writeByte('}')
+	}
+
+	t := di.d.nextCBORType()
+	switch t {
+	case cborTypePositiveInt:
+		_, _, val := di.d.getHead()
+		return di.writeString(strconv.FormatUint(val, 10))
+
+	case cborTypeNegativeInt:
+		_, _, val := di.d.getHead()
+		if val > math.MaxInt64 {
+			// CBOR negative integer overflows int64, use big.Int to store value.
+			bi := new(big.Int)
+			bi.SetUint64(val)
+			bi.Add(bi, big.NewInt(1))
+			bi.Neg(bi)
+			return di.writeString(bi.String())
+		}
+
+		nValue := int64(-1) ^ int64(val)
+		return di.writeString(strconv.FormatInt(nValue, 10))
+
+	case cborTypeByteString:
+		b := di.d.parseByteString()
+		return di.encodeByteString(b)
+
+	case cborTypeTextString:
+		b, err := di.d.parseTextString()
+		if err != nil {
+			return err
+		}
+		return di.encodeTextString(string(b), '"')
+
+	case cborTypeArray:
+		_, _, val := di.d.getHead()
+		count := int(val)
+		if err := di.writeByte('['); err != nil {
+			return err
+		}
+
+		for i := 0; i < count; i++ {
+			if i > 0 {
+				if err := di.writeString(", "); err != nil {
+					return err
+				}
+			}
+			if err := di.item(); err != nil {
+				return err
+			}
+		}
+		return di.writeByte(']')
+
+	case cborTypeMap:
+		_, _, val := di.d.getHead()
+		count := int(val)
+		if err := di.writeByte('{'); err != nil {
+			return err
+		}
+
+		for i := 0; i < count; i++ {
+			if i > 0 {
+				if err := di.writeString(", "); err != nil {
+					return err
+				}
+			}
+			// key
+			if err := di.item(); err != nil {
+				return err
+			}
+			if err := di.writeString(": "); err != nil {
+				return err
+			}
+			// value
+			if err := di.item(); err != nil {
+				return err
+			}
+		}
+		return di.writeByte('}')
+
+	case cborTypeTag:
+		_, _, tagNum := di.d.getHead()
+		switch tagNum {
+		case 2:
+			b := di.d.parseByteString()
+			bi := new(big.Int).SetBytes(b)
+			return di.writeString(bi.String())
+
+		case 3:
+			b := di.d.parseByteString()
```

Comment:
Need to check tag content type to avoid runtime error.

```suggestion
			if di.d.nextCBORType() != cborTypeByteString {
				return errors.New("cbor: tag number 3 must be followed by byte string, got " + t.String())
			}
			b := di.d.parseByteString()
```